test/python/transpiler/test_template_matching.py

```diff
@@ -353,9 +354,24 @@ def template():
         pass_ = TemplateOptimization(template_list=[template()])
         circuit_out = PassManager(pass_).run(circuit_in)
 
-        # This template will not fully match as long as gates with parameters do not
-        # commute with any other gates in the DAG dependency.
-        self.assertEqual(circuit_out.count_ops().get("cx", 0), 2)
+        self.assertEqual(circuit_out.count_ops().get("cx", 0), 0)
```

Comment:
I know this was pre-existing, but it'd be good to make this a test of the entire structure of the circuit, not just an op-count - as I understand it, this test should now return an empty circuit, so we should test that.

(As an aside: it maybe would be good to update the test slightly so it doesn't return a fully empty circuit, because a buggy feature completely wiping out all data in the circuit seems like a plausible failure mode.)

Reply:
I was attempting to improve this test by using `numpy.testing.assert_allclose` and passing it the `circuit_out` with bound parameters through the `Operator` class from `qiskit.quantum_info`, and ran right into an [issue I raised](https://github.com/Qiskit/qiskit-terra/issues/6633) dealing with `Parameter`s not being removed from the parameter table. Since this is not yet fixed, I do not think there is a good way to solve this currently.

Perhaps we could move this point to another Issue?

Reply:
I looked more, and this test is vitally important, and this PR breaks it as it is.  The output of the test _should_ contain 2 CX gates.  If you run this test with the changes in this PR, you get
```python
In [4]: circuit_out.draw()
Out[4]:
     ┌────┐
q_0: ┤0   ├─────────────
     │  P │┌───────────┐
q_1: ┤1   ├┤ P(-1.0*β) ├
     └────┘└───────────┘
```
Note that it's replaced the two CXs with two gates whose parameters don't mean anything.  This is a failure; the pass should not have replaced anything, even with your changes.  If the test had stronger assertions (like I was suggesting in the previous comment), it would have been clearer that this was a true test failure.

Fortunately, now that I understand this, we can hugely simplify the requirements on `UnitaryGate`; we can rewrite the test to avoid it by just using a gate that can handle `Parameter` in its inverse.  Here's a clearer version of the test that's exactly equivalent:

```python
    def test_optimizer_does_not_replace_unbound_partial_match(self):
        beta = Parameter("β")
        template = QuantumCircuit(2)
        template.cx(1, 0)
        template.cx(1, 0)
        template.p(beta, 1)
        template.cu(0, 0, 0, -beta, 0, 1)

        circuit_in = QuantumCircuit(2)
        circuit_in.cx(1, 0)
        circuit_in.cx(1, 0)
        pass_ = TemplateOptimization(
            template_list=[template],
            user_cost_dict={"cx": 6, "p": 0, "cu": 8},
        )

        circuit_out = PassManager(pass_).run(circuit_in)

        # The template optimisation should not have replaced anything, because
        # that would require it to leave dummy parameters in place without
        # binding them.
        self.assertEqual(circuit_in, circuit_out)
```

If we change the test to that, it lets us revert all the changes to `UnitaryGate` that I now see are unrelated (sorry again about that), and the test is both clearer and stricter.

The trick, though, is that I think to achieve your goals in the pass, you're going to need to update your logic a bit.  You need to keep track of which parameters are dummy parameters used in the template, and which are parameters that come from the circuit.  For example, in this test, the `beta` should not be allowed to be output by the transpiler pass, but in the test you added after this one, the `phi` _should_ be output (as you have it).

There's more nuance than just that as well, though, because the pass needs to handle the case that a template uses a parameter with the same name as an input circuit without breaking.  You can likely handle that bit by re-binding the variables in a template to auto-generated unique names if you detect issues.

Reply:
That's a great point to eliminate the `UnitaryGate` from the unit test entirely (I implemented the above, and also modified `test_parametric_template` similarly). The `test_unbound_parameters_in_rzx_template` test is now working so that `circuit_in` and `circuit_out` are verified to not be equal (because the template optimization has changed the circuit), yet the operators are equivalent after binding `phi` to a `float`.

In order to guarantee template optimization does *not increase* the number of `Parameter`s in a circuit, there is a new private method `_incr_num_parameters` in `template_substitution.py` that counts the number of `Parameter`s in the `template` and in the `circuit` and returns true if the former is greater than the latter. If this is true, logic inside `_substitution` will continue out of the method, preventing the template from being substituted.

As for `Parameter`s in circuits and templates that have the same name, I'm not sure I know how to auto-generate unique names. Trying this out on the test notebook, all the `template`s I use are of the `rzx` variety, and the specific `Parameter` used is always
```
theta = Parameter("ϴ") # used in all rzx templates
```
which does not match with
```
theta = Parameter('θ') # typed from my "Greek" keyboard
```
or
```
theta = Parameter('$\\theta$') # maybe the safest, LaTeX version
```
so it seems that this is not much of a problem currently. (Of course template optimization fails if I use the exact form of `theta` from the template).